### PR TITLE
[BEAM-6783] byte[] breaks in BeamSQL codegen

### PR DESCRIPTION
**Pull request by kanterov**
Fix casting in BeamCalcRel.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by kanterov** [comment]
Run JavaPortabilityApi PreCommit

**Comment by kanterov** [comment]
R: @apilloud @amaliujia

**Comment by amaliujia** [comment]
LGTM

**Comment by kanterov** [comment]
@amaliujia @apilloud in JIRA you mentioned UDFs, is there anything else that needs to be fixed or I can close the ticket?

**Comment by kanterov** [comment]
Tests failed, it seems yes :)

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 1]
```diff
@@ -411,6 +414,8 @@ public Expression field(BlockBuilder list, int index, Type storageType) {
                 Expressions.equal(field, Expressions.constant(null)),
                 Expressions.constant(null),
                 Expressions.call(WrappedList.class, "of", field));
+      } else if (fromType.getTypeName() == TypeName.BYTES && storageType == ByteString.class) {
```
Comment (kanterov) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: Had to add "storageType == ByteString.class", "InputGetterImpl" is used to convert Beam schema types to UDF parameters, that we don't always want for our own UDFs.

**Comment by kanterov** [comment]
Run JavaPortabilityApi PreCommit

**Comment by kanterov** [comment]
Run JavaPortabilityApi PreCommit

**Comment by kanterov** [comment]
@amaliujia @apilloud I had to adjust UDF code to handle BYTES. It turns out there is an issue with how we input and output nullable parameters from UDFs. I suspect there are more things broken in this area. I want to merge this as-is and other existing issues later.

**Comment by kanterov** [comment]
@amaliujia @apilloud I did one more commit to annotate builtin string UDFs with Strict. It significantly simplifies the code and fixes a few of existing TODOs. However, tests will pass without this change as well.

**Comment by amaliujia** [comment]
@kanterov Could you explain what does `@Strict` annotation do? I had a hard time to find good reference for this annotation.

**Comment by kanterov** [comment]
@amaliujia there is no documentation, only source code. Calcite infers NullPolicy based on annotations: https://github.com/apache/beam/blob/f3a2f96c05552c420ca351b77abc03ad577604d3/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/ScalarFunctionImpl.java#L188-L197 and NullPolicy declares how to deal with nulls: https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/adapter/enumerable/NullPolicy.java

**Comment by amaliujia** [comment]
Thanks for your pointers!

I am thinking maybe we should use `@ANY` policy? `@Strict` says return null **if and only if** one of the arguments are null, but read like it could fail to return null if more than one argument is null?

**Comment by kanterov** [comment]
@amaliujia good point, I was looking into Calcite code and didn't find a definite answer. In my understanding, Strict documentation should say "returns null iff (if and only if) at least one argument is null", meaning that UDF will never accept or return null, and wrapping implementor will handle that.


https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java#L287

**Comment by amaliujia** [comment]
LGTM

Agree. I did some test and seems `Strict` is doing `at least one argument is null`. I also didn't find an annotation of `@Any`.

**Comment by kanterov** [comment]
@amaliujia @apilloud agree, thanks for the review!

**Pull request merged**